src/coreclr/jit/morph.cpp

```diff
@@ -12940,56 +12950,19 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
         return node;
     }
 
-    switch (node->GetHWIntrinsicId())
+    simd32_t simd32Val = {};
+
+    if (GenTreeVecCon::IsHWIntrinsicCreateConstant(node, simd32Val))
     {
-        case NI_Vector128_Create:
-#if defined(TARGET_XARCH)
-        case NI_Vector256_Create:
-#elif defined(TARGET_ARM64)
-        case NI_Vector64_Create:
-#endif
-        {
-            bool hwAllArgsAreConstZero = true;
-            for (GenTree* arg : node->Operands())
-            {
-                if (!arg->IsIntegralConst(0) && !arg->IsFloatPositiveZero())
-                {
-                    hwAllArgsAreConstZero = false;
-                    break;
-                }
-            }
+        GenTreeVecCon* vecCon = gtNewVconNode(node->TypeGet());
 
-            if (hwAllArgsAreConstZero)
-            {
-                switch (node->GetHWIntrinsicId())
-                {
-                    case NI_Vector128_Create:
-                    {
-                        node->ResetHWIntrinsicId(NI_Vector128_get_Zero);
-                        break;
-                    }
-#if defined(TARGET_XARCH)
-                    case NI_Vector256_Create:
-                    {
-                        node->ResetHWIntrinsicId(NI_Vector256_get_Zero);
-                        break;
-                    }
-#elif defined(TARGET_ARM64)
-                    case NI_Vector64_Create:
-                    {
-                        node->ResetHWIntrinsicId(NI_Vector64_get_Zero);
-                        break;
-                    }
-#endif
-                    default:
-                        unreached();
-                }
-            }
-            break;
+        for (GenTree* arg : node->Operands())
+        {
+            DEBUG_DESTROY_NODE(arg);
         }
 
-        default:
-            break;
+        vecCon->gtSimd32Val = simd32Val;
+        return fgMorphTree(vecCon);
```

Comment:
This method is called in post-order and so should not recursively call `fgMorphTree`.

Reply:
We get an assert that morph wasn't called on the node if we don't. Should we just be manually marking it as having already been morphed?

Reply:
> Should we just be manually marking it as having already been morphed?

Yep, that's the pattern we use for similar situations:

https://github.com/dotnet/runtime/blob/9158543338acd47019ffb6ec634a273a9e28cb99/src/coreclr/jit/morph.cpp#L4726-L4731

https://github.com/dotnet/runtime/blob/9158543338acd47019ffb6ec634a273a9e28cb99/src/coreclr/jit/morph.cpp#L10568-L10572

Reply:
Fixed.